### ISSUE TITLE
If a spawned sub-task awaits and never returns, handle will not join.

### ISSUE THREAD

**Issue by Firstyear** [QnnOkabayashi/tracing-forest#4]
If we have a set of socket listeners such as what is in tide

```
tokio::spawn(async move {
    let _ = app.listen(listener).await;
})
```

Then in `tracing-forest/src/builder.rs` in on_future, `handle.await.expect("failed closing the writing task");` will never return as there are still remaining processors which prevents the tx channel from being dropped (I assume).

I'm not sure if this is a bug in tide ( no way to signal a shutdown ) or a bug in tracing-forest (no timeout on the handle.await).

**Comment by QnnOkabayashi** [comment]
I don't think this should be an issue, because `on_future` uses [`tokio::spawn`](https://docs.rs/tokio/latest/tokio/fn.spawn.html) to create a concurrent logging task. This means that it should be logging even if the `Future` passed to `on_future` runs indefinitely.

**Comment by QnnOkabayashi** [comment]
I think what's happening here is that whatever process `app.listen(listener)` is doing is wrapped in a span. Since `tracing-forest` only prints when a root span is closed, somehow you're leaving the scope of the `Future` passed to `on_future` before exiting the root span in that spawned task, meaning that the subscriber doesn't have the opportunity to print all the trace data it collects before it's dropped.

I think you could resolve this by using a `oneshot` to send a graceful shutdown signal to your `tide` task to close out of any spans, so that `tracing-subscriber` knows to print everything. If this is actually the issue, we should definitely mention something like this in the docs in a highly visible place.

**Comment by QnnOkabayashi** [comment]
If you wanted to do a ctrl c shutdown, you can use `tokio::select!` to have 
```rust
// setup
tokio::select! {
    Ok(()) = tokio::signal::ctrl_c() => {}
    _ = tokio::spawn(app.listen(listener)) => {} // or `app.listen(listener)` if you want don't want it to dangle
}
// shutdown
```

**Comment by Firstyear** [comment]
Right, this fixes it, but I think it would be a good thing to have in the docs if someone chooses to build their own asyncy bits.

Saying this, worth also saying that this DOES reproduce with the attributes meaning that it's very very very easy to shoot yourself with this and cause the proc to hang.

**Comment by QnnOkabayashi** [comment]
Resolved in #9, closing this issue.
